### ISSUE TITLE
README.md claims no native code is used but...

### ISSUE THREAD

**Issue by cowwoc** [mabe02/lanterna#255]
I see that the Windows implementation relies on JNA which uses native code. 

Consider removing the native code in light of http://superuser.com/a/1105718/57662 or modifying the readme to explain that native code is used under Windows.

**Comment by dant3** [comment]
Firstly JNA is not native code - it is Java code which calls in some
existing native code, so library itself is pure Java. Secondarily Windows
10 market share is 20% compared to 50% of Windows 7. Can't see reason to
drop 50% in favour of 20% just to get rid of little piece of native calls
which are well isolated, and only used on a single platform.

**Comment by cowwoc** [comment]
@dant3 I get the fact that this library does not contain a native component, but the very fact that you rely upon a library which does (JNA) means that users have to worry about deploying different native libraries per platform. It doesn't matter whether the native code comes your library or your transitive dependencies, the deployment headache follows. I hope this clarifies what I meant.

Good point regarding the market share of Windows 7. I was just attempting to inform you that Windows 10 now supports basic ANSI codes in case you were unaware.

**Comment by dant3** [comment]
The functionality it uses via JNA is built into your platform, you have no
dependencies but the platform itself. On Linux it also uses TTY which is
native, but it's already there, so nothing to worry about.

**Comment by cowwoc** [comment]
I understand. What I meant is that you have to ship the different JNA native libraries for each platform and at launch time it unpacks the DLLs/SOs from the JAR into some temporary directory, and so on. This deployment process is not perfect and I've seen people run into problems (failure to unpack the files, files never getting deleted, etc). Anyway, I think that we're on the same page now so you can go ahead and close this issue. Thank you.

**Comment by avl42** [comment]
that's exactly what you do NOT need to do!

if a user of your application runs on windows console (within a cmd.exe
window), then he will already have the appropriate kernel.dll, if he runs
on windows with swing terminal, or on unix (either tty or swing) then the
jna code parts shouldn't even be used.

**Comment by cowwoc** [comment]
@avl42 It sounds like we have a misunderstanding here.

In the scenario where a user is running inside cmd.exe on Windows, you will have to use JNA. Do you agree?

The minute you are using JNA you are actually using JNI under the hood. The minute you are doing that, you have to bundle a different native library per platform then at runtime you have to unpack and load said library. Do you agree?

**Comment by avl42** [comment]
@cowwoc I don't agree.   Java isn't like compiled C.  In Java classes are only loaded on demand, and as long as execution doesn't hit(*) the specific lanterna class for console access on Windows, the JVM itself doesn't even "know" that your application could potentially use JNA.

(*) "hit" means: call a static method of that class or access a static field of it or make a new object of that class, and a couple more, but definitely excluding merely having the class name show up in the DefaultTerminalFactory for code paths not taken.

Just wondering: have you actually tried packaging just your application & lanterna.jar (and no .dll or .so or .dl or whatsoever) and run it on various platforms? Did you see any problems?

I must admit, I haven't tried it - as our customer still runs the beta2 (which didn't have the jna stuff, yet, and they don't run it on Windows, anyway). I'm still pretty sure about it.

**Comment by cowwoc** [comment]
@avl42 It sounds like we are talking past each other a little bit so allow me to clarify.

I never claimed that lanterna was loading native libraries on platforms that didn't need it. I'm saying that when the native code *is* needed (e.g. using cmd.exe under Windows) JNA gets used, which means that JNI gets used, which means that the README.md's implicit assertions that this library is 100% pure Java code is not strictly true.

To me "100% pure Java" means that I don't have to worry about the complications of bundling/deploying native libraries. It doesn't matter whether Lanterna contains native libraries directly or its transitive dependencies do. As soon as there is at least one native library at play the user of this library pays the price.

**Comment by avl42** [comment]
If shipping native code (as part of another third party library) really **is** a problem, then unfortunately your only alternative is to declare Windows _console_ mode as _unsupported_ and ship your application and lanterna.jar _without_ any JNA library.  I highly doubt that any "really 100% pure Java" library could support this target at all.

PS: I'm sorry, I didn't understand the "transitive dependencies" part earlier. To be honest, I thought that JNA was part of the Java Standard Library, but it really isn't. Eclipse never complained about the new stuff, but that was only because I had forgotten to add the native-integration hierarchy to the build-path...

**Comment by cowwoc** [comment]
Sounds like we're in agreement ;)

Thank you both for the clarification. This is a cool library and I'm going to keep my eye on it in the future.

**Comment by mabe02** [comment]
O_O

**Comment by mabe02** [comment]
While the JNA-dependent component is part of the same repository, it's not really part of the library itself. The output of the main maven project is a jar file that doesn't rely on JNA or any native library code, but unfortunately it doesn't work with the default Windows console. I've tried to get Cygwin working with lanterna, but it's really tricky when you run a win32 executable that isn't linked against any of the cygwin libraries inside the cygwin environment.
So, just for fun I put together a support library that depends on jna that adds experimental support. I got it working with Windows 7 but when I tried it a while back with Windows 10 it didn't work so I'm not sure how useful it really is at the moment.

Anyhow, I don't consider this support library to be really a part of lanterna itself. Maybe I should have made a separate git repository just to clarify this?
